boundary-data-merge.py

```diff
@@ -0,0 +1,97 @@
+#!/bin/env python3
+
+import csv
+import json
+import os
+import shutil
+import tempfile
+
+import requests
+
+wikidata_ids = set()
+
+# We'll use this to recreate the original line-endings in CSV files. Can't
+# easily get it on the second (write) pass as we need it before the CSV reader
+# has read any data from the file
+newlines = {}
+
+if os.path.isdir(os.path.join('boundaries', 'build')):
+    boundaries_dir = os.path.join('boundaries', 'build')
+else:
+    boundaries_dir = 'boundaries'
+
+boundaries_index_fn = os.path.join(boundaries_dir, 'index.json')
+with open(boundaries_index_fn) as f:
+    boundaries_index = json.load(f)
+directories = {entry['directory'] for entry in boundaries_index}
+
+# Gather Wikidata IDs from CSV files
+for directory in directories:
+    csv_fn = os.path.join(boundaries_dir, directory, directory + '.csv')
+    with open(csv_fn, newline='') as f:
+        data = csv.DictReader(f)
+        wikidata_ids |= {row['WIKIDATA'] for row in data}
+        newlines[csv_fn] = f.newlines
+
+# And from associated positions in the boundary index
+for entry in boundaries_index:
+    for association in entry['associations']:
+        wikidata_ids.add(association['position_item_id'])
+
+# Query for any replacement IDs
+query = """
+SELECT ?old ?new WHERE {
+  VALUES ?old { %s }
+  ?old owl:sameAs ?new
+}
+""" % ' '.join('wd:' + wikidata_id for wikidata_id in wikidata_ids)
```

Comment:
BTW, I have a very similar helper to this in EveryPolitican, and had to restrict it to checking 500 ids at a time, as the query service returns a `414 URI too long` error for much more than that.

No idea if that's likely to be an issue here, and I didn't really spend much time investigating whether there was any way of working around that, as it was easy enough to loop over them with a `each_slice(500).flat_map` (in Ruby), but if you're potentially going to be dealing with quite large numbers of IDs, it might be worth testing what your equivalent limits might be. 

(And FWIW, my version also does the equivalent of

```sparql
SELECT
    (STRAFTER(STR(?old), STR(wd:)) AS ?old)
    (STRAFTER(STR(?new), STR(wd:)) AS ?new)
```

to avoid needing to split on the returned values later)

Reply:
Cool. This is POSTing the query, not GETing it, so URI length shouldn't be a problem. Also didn't know about `STRAFTER`; I've always done `SUBSTR(STR(?old), 32)`, which is a tad opaque.

Reply:
Yes, my version also uses `POST`, so I assumed it was triggering a configured maxlength for variables (AIUI there's no conceptual limit on those, but most servers impose one). Does your version successfully run even with 1000+ ids? If so, there must have been something else odd going on with mine.

Reply:
I've just tried it with 55500 by changing the script to submit for `list(wikidata_ids) * 100`, and it took a while, but it did return successfully. I'd expect you to get a `413 Payload Too Large` for too big a request being POSTed, instead of a 414 for the URI.